include/flatbuffers/grpc.h

```diff
@@ -23,50 +23,271 @@
+namespace detail {
+struct SliceAllocatorMember {
+  SliceAllocator slice_allocator_;
+};
+}
+
+// MessageBuilder is a gRPC-specific FlatBufferBuilder that uses SliceAllocator
+// to allocate gRPC buffers.
+class MessageBuilder : private detail::SliceAllocatorMember,
+                       public FlatBufferBuilder {
+ public:
+  explicit inline MessageBuilder(uoffset_t initial_size = 1024)
+    : FlatBufferBuilder(initial_size, &slice_allocator_, false),
+      initial_size_(initial_size) {}
+
+  MessageBuilder(const MessageBuilder &other) = delete;
+  MessageBuilder &operator=(const MessageBuilder &other) = delete;
+
+  inline ~MessageBuilder() {}
+
+  // GetMessage extracts the subslice of the buffer corresponding to the
+  // flatbuffers-encoded region and wraps it in a `Message<T>` to handle buffer
+  // ownership.
+  template <class T>
+  inline Message<T> GetMessage() {
+    vector_downward &vec = GetVectorDownward();
+    uint8_t *buf_data = vec.buf();     // pointer to memory
+    size_t buf_size = vec.capacity();  // size of memory
+    uint8_t *msg_data = vec.data();    // pointer to msg
+    size_t msg_size = vec.size();      // size of msg
+
+    // Do some sanity checks on data/size
+    assert(msg_data >= msg_data);
+    assert(msg_data + msg_size <= buf_data + buf_size);
+
+    // Calculate offsets from the buffer start
+    size_t begin = msg_data - buf_data;
+    size_t end = begin + msg_size;
+
+    // Get the slice we are working with (no refcount change)
+    grpc_slice slice = slice_allocator_.get_slice(buf_data, buf_size);
+
+    // Extract a subslice of the existing slice (increment refcount)
+    grpc_slice subslice = grpc_slice_sub(slice, begin, end);
+
+    // Wrap the subslice in a `Message<T>` (steal ref since incremented above)
+    Message<T> msg(subslice, STEAL_REF);
+
+    return msg;
+  }
+
+  template <class T>
+  inline Message<T> ReleaseMessage() {
+    Message<T> msg = GetMessage<T>();
+    Reset(initial_size_, &slice_allocator_, false);
+    return msg;
+  }
+
+ private:
+  // SliceAllocator slice_allocator_;  // part of SliceAllocatorMember
+  uoffset_t initial_size_;
+};
+
 template <class T>
-class SerializationTraits<T, typename std::enable_if<std::is_base_of<
-                                 flatbuffers::BufferRefBase, T>::value>::type> {
+class MessageVerifier {
  public:
-  // The type we're passing here is a BufferRef, which is already serialized
-  // FlatBuffer data, which then gets passed to GRPC.
-  static grpc::Status Serialize(const T& msg,
-                                grpc_byte_buffer **buffer,
-                                bool *own_buffer) {
-    // TODO(wvo): make this work without copying.
-    auto slice = gpr_slice_from_copied_buffer(
-                   reinterpret_cast<const char *>(msg.buf), msg.len);
-    *buffer = grpc_raw_byte_buffer_create(&slice, 1);
-    grpc_slice_unref(slice);
+  static inline ::grpc::Status Verify(const Message<T> &msg) {
```

Comment:
Ah ok, so it is overrideable. So now we have both a pre-processor flag and an overrideable template, just to save the user from calling a function manually.. this may be a bit much. If we have `Verify` in all examples and documentation, that should be enough for users to consider if they want it.